fdbrpc/include/fdbrpc/AsyncFileWriteChecker.actor.h

```diff
@@ -0,0 +1,381 @@
+			self->updateChecksumHistory(false, offset, *length, (uint8_t*)data);
+			return r;
+		});
+	}
+
+	Future<Void> write(void const* data, int length, int64_t offset) override {
+		auto pages = updateChecksumHistory(true, offset, length, (uint8_t*)data);
+		auto self = Reference<AsyncFileWriteChecker>::addRef(this);
+		return map(m_f->write(data, length, offset), [self, pages](Void r) {
+			for (uint32_t page : pages) {
+				self->writing.erase(page);
+			}
+			return r;
+		});
+	}
+
+	Future<Void> truncate(int64_t size) override {
+		auto self = Reference<AsyncFileWriteChecker>::addRef(this);
+		return map(m_f->truncate(size), [self, size](Void r) {
+			int maxFullPage = size / checksumHistoryPageSize;
+			int oldSize = self->lru.size();
+			self->lru.truncate(maxFullPage);
+			checksumHistoryBudget.get() += oldSize - self->lru.size();
+			return r;
+		});
+	}
+
+	Future<Void> sync() override {
+		auto self = Reference<AsyncFileWriteChecker>::addRef(this);
+		return map(m_f->sync(), [self](Void r) {
+			self->syncedTime = AsyncFileWriteChecker::transformTime(now());
+			return r;
+		});
+	}
+
+	Future<Void> flush() override { return m_f->flush(); }
+	Future<int64_t> size() const override { return m_f->size(); }
+	std::string getFilename() const override { return m_f->getFilename(); }
+	void releaseZeroCopy(void* data, int length, int64_t offset) override {
+		return m_f->releaseZeroCopy(data, length, offset);
+	}
+	int64_t debugFD() const override { return m_f->debugFD(); }
+
+	struct WriteInfo {
+		WriteInfo() : checksum(0), timestamp(0) {}
+		uint32_t checksum;
+		uint32_t timestamp; // keep a precision of ms
+	};
+
+	struct node {
+		uint32_t page;
+		WriteInfo writeInfo;
+		node *next, *prev;
+		node(uint32_t _page, WriteInfo _writeInfo) {
+			page = _page;
+			writeInfo = _writeInfo;
+			page = _page;
+			next = NULL;
+			prev = NULL;
+		}
+	};
+
+	class LRU {
+	private:
+		node* start;
+		node* end;
+		std::unordered_map<uint32_t, node*> m;
+		std::string fileName;
+		int maxFullPage;
+
+		void insertHead(node* n) {
+			n->next = start->next;
+			start->next->prev = n;
+			n->prev = start;
+			start->next = n;
+		}
+
+		void removeFromList(node* n) {
+			n->prev->next = n->next;
+			n->next->prev = n->prev;
+		}
+
+	public:
+		LRU(std::string _fileName) {
+			fileName = _fileName;
+			maxFullPage = 0;
+			start = new node(0, WriteInfo());
+			end = new node(0, WriteInfo());
+			start->next = end;
+			end->prev = start;
+		}
+
+		void update(uint32_t page, WriteInfo writeInfo) {
+			if (m.find(page) != m.end()) {
+				node* n = m[page];
+				removeFromList(n);
+				insertHead(n);
+				n->writeInfo = writeInfo;
+				return;
+			}
+			node* n = new node(page, writeInfo);
+			insertHead(n);
+			m[page] = n;
+			if (page >= maxFullPage) {
+				maxFullPage = page + 1;
+			}
+		}
+
+		int maxPage() { return maxFullPage; }
+
+		void truncate(int newMaxFullPage) {
+			for (int i = newMaxFullPage; i < maxFullPage; i++) {
```

Comment:
if the `maxFullPage - newMaxFullPage` is large, i.e., >> total number of entries in the LRU, then this is not efficient (and iterating through linked list is faster). You can benchmark the CPU time similar to the ones in `flowbench` directory. We don't want this to cause slow tasks.

Reply:
thanks for the comments, they all make sense, for this particular one, yeah I understand your concern, I thought about it too, however the linked list is sorted in the order of the most recent use, so it seems there is no easy way.  

an `unordered_map` has been used here, so its ::find and ::erase is O(1). So I believe its complexity is the same as iterating the list, i.e. O(# of elements to remove).

Please let me know if I understand it correctly, we can talk offline. @jzhou77

Reply:
> an unordered_map has been used here, so its ::find and ::erase is O(1). So I believe its complexity is the same as iterating the list, i.e. O(# of elements to remove).

Even though ::find is O(1), doing `maxFullPage - newMaxFullPage` times can be much larger than `O(# of elements to remove)`.  That's why I am suggesting to benchmark the CPU time. If this can be an issue with benchmark results, we should think about optimize for such cases, e.g., iterating the map instead of iterating page numbers.

Reply:
ah i see, makes total sense, for instance, if this `maxFullPage - newMaxFullPage` is larger than the map, it is certainly better to just iterate the map.